### PR TITLE
kube client cert auth in the proxy

### PR DESCRIPTION
**Pull request by csams**
## Summary
This PR wires up kube machinery for client cert auth in the front proxy.

Signed-off-by: Christopher Sams <csams@redhat.com>

### CODE REVIEW

File: cmd/kcp-front-proxy/options/options.go [hunk 1]
```diff
@@ -17,27 +17,62 @@ limitations under the License.
...
 
 	proxyoptions "github.com/kcp-dev/kcp/pkg/proxy/options"
 )
 
-type Options struct {
-	SecureServing *apiserveroptions.SecureServingOptionsWithLoopback
-	Proxy         proxyoptions.Options
+// ClientCertOptions wraps ClientCertAuthenticationOptions so we don't pull in
+// more auth machinery than we need with DelegatingAuthenticationOptions
+type ClientCertOptions struct {
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/options.go:35]: call this Authentication

File: cmd/kcp-front-proxy/options/options.go [hunk 2]
```diff
@@ -17,27 +17,62 @@ limitations under the License.
...
+	Options apiserveroptions.ClientCertAuthenticationOptions
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/options.go:36]: call this ClientCert

File: cmd/kcp-front-proxy/options/options.go (hunk 1 above)
Comment (sttts) [reply cmd/kcp-front-proxy/options/options.go:35]: and move into authentication.go

File: cmd/kcp-front-proxy/options/options.go [hunk 3]
```diff
@@ -17,27 +17,62 @@ limitations under the License.
...
+}
+
+// ApplyTo sets up the x509 Authenticator if the client-ca-file option was passed
+func (c *ClientCertOptions) ApplyTo(authenticationInfo *genericapiserver.AuthenticationInfo, servingInfo *genericapiserver.SecureServingInfo) error {
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/options.go:40]: move to authentication.go

File: cmd/kcp-front-proxy/options/options.go [hunk 4]
```diff
@@ -17,27 +17,62 @@ limitations under the License.
...
+// Validate just completes the options pattern. Returns nil.
+func (c *ClientCertOptions) Validate() []error {
+	return nil
+}
+
+type Options struct {
+	SecureServing  apiserveroptions.SecureServingOptionsWithLoopback
+	Authentication ClientCertOptions
+	Proxy          proxyoptions.Options
+	RootDirectory  string
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/options.go:68]: empty line above

File: cmd/kcp-front-proxy/options/options.go [hunk 5]
```diff
@@ -17,27 +17,62 @@ limitations under the License.
...
 }
 
 func NewOptions() *Options {
 	o := &Options{
-		SecureServing: apiserveroptions.NewSecureServingOptions().WithLoopback(),
-		Proxy:         *proxyoptions.NewOptions(),
+		SecureServing:  *apiserveroptions.NewSecureServingOptions().WithLoopback(),
+		Authentication: ClientCertOptions{},
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/options.go:74]: NewAuthentication()

File: cmd/kcp-front-proxy/options/options.go [hunk 6]
```diff
@@ -17,27 +17,62 @@ limitations under the License.
...
 	apiserveroptions "k8s.io/apiserver/pkg/server/options"
 
 	proxyoptions "github.com/kcp-dev/kcp/pkg/proxy/options"
 )
 
-type Options struct {
-	SecureServing *apiserveroptions.SecureServingOptionsWithLoopback
-	Proxy         proxyoptions.Options
+// ClientCertOptions wraps ClientCertAuthenticationOptions so we don't pull in
+// more auth machinery than we need with DelegatingAuthenticationOptions
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/options.go:34]: just to be clear: we definitely do not want full authentiaction, e.g. oidc or similar here. JWT should just passed on.

File: cmd/kcp-front-proxy/main.go [hunk 7]
```diff
@@ -72,17 +78,33 @@ routed based on paths.`,
...
+				return err
+			}
+
+			// If an authenticator was set, wrap the handler with one that uses it
+			if authenticationInfo.Authenticator != nil {
+				scheme := runtime.NewScheme()
+				metav1.AddToGroupVersion(scheme, schema.GroupVersion{Group: "", Version: "v1"})
+				codecs := serializer.NewCodecFactory(scheme)
+
+				failedHandler := genericapifilters.Unauthorized(codecs)
```
Comment (sttts) [inline cmd/kcp-front-proxy/main.go:103]: is the failedHandler called for everything other than client certs? That wouldn't make much sense.

Comment (csams) [reply cmd/kcp-front-proxy/main.go]: Yeah, I realized that and wrote a different handler.

File: cmd/kcp-front-proxy/options/options.go (hunk 6 above)
Comment (csams) [reply cmd/kcp-front-proxy/options/options.go]: Right - I looked at delegating auth first and realized it pulled in too much. So I decided to use the ClientCertAuthenticationOptions instead.

File: cmd/kcp-front-proxy/options/authentication.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+// NewAuthentication creates a default Authentication
+func NewAuthentication() *Authentication {
+	return &Authentication{}
+}
+
+// ApplyTo sets up the x509 Authenticator if the client-ca-file option was passed
+func (c *Authentication) ApplyTo(authenticationInfo *genericapiserver.AuthenticationInfo, servingInfo *genericapiserver.SecureServingInfo) error {
+	clientCAProvider, err := c.ClientCert.GetClientCAContentProvider()
+	if err != nil {
+		return fmt.Errorf("unable to load client CA provider: %w", err)
```
Comment (sttts) [inline cmd/kcp-front-proxy/options/authentication.go:44]: does this survice running it witht and without the cert? Does this wire the dynamic CA cert loading?

Comment (sttts) [inline cmd/kcp-front-proxy/options/authentication.go:44]: does this survice running it witht and without the cert? Does this wire the dynamic CA cert loading?

**Review by sttts — APPROVED** [review]

**Pull request merged**